### PR TITLE
Sprint 11 Prep Task 5: Prototype Factoring Algorithms

### PR DESCRIPTION
**Pull request by jeffreyhorn**
## Summary

Prototype distribution cancellation and multi-term factoring algorithms for aggressive simplification, **exceeding all performance and effectiveness targets**.

## Key Results

- ✅ **39.2% average reduction** (nearly 2× the 20% target)
- ✅ **Execution time 57× faster** than 1ms threshold (0.0175ms max)
- ✅ **All 6 benchmark cases pass** performance and effectiveness targets
- ✅ **7/7 tests passing** with comprehensive edge case coverage

## Unknowns Verified

- ✅ **Unknown 1.1:** Common factor detection via set-based AST structural equality
- ✅ **Unknown 1.7:** Factoring performance exceeds all targets

## Files Created

- `prototypes/aggressive_simplification/factoring_prototype.py` - Core algorithms (~380 lines)
- `prototypes/aggressive_simplification/test_factoring.py` - Test suite (7/7 passing, ~200 lines)
- `prototypes/aggressive_simplification/benchmark_factoring.py` - Performance benchmarks (~150 lines)
- `docs/planning/EPIC_2/SPRINT_11/factoring_prototype_results.md` - Comprehensive results document

## Files Updated

- `docs/planning/EPIC_2/SPRINT_11/KNOWN_UNKNOWNS.md` - Unknowns 1.1, 1.7 verified
- `docs/planning/EPIC_2/SPRINT_11/PREP_PLAN.md` - Task 5 marked complete
- `CHANGELOG.md` - Task 5 entry added

## Performance Benchmarks

| Test Case | Before | After | Reduction | Time (ms) | Status |
|-----------|--------|-------|-----------|-----------|--------|
| Simple 2-term (x*y + x*z) | 3 ops | 2 ops | 33.3% | 0.0098 | ✅ |
| Three terms (x*y + x*z + x*w) | 5 ops | 3 ops | 40.0% | 0.0143 | ✅ |
| Multiple common (2*x*y + 2*x*z) | 5 ops | 3 ops | 40.0% | 0.0138 | ✅ |
| PROJECT_PLAN.md example | 5 ops | 3 ops | 40.0% | 0.0141 | ✅ |
| Four terms | 7 ops | 4 ops | 42.9% | 0.0175 | ✅ |
| No common factors | 3 ops | 3 ops | 0.0% | 0.0046 | ✅ |

## Recommendations for Sprint 11

1. ✅ Integrate `factor_common_terms()` into simplification pipeline
2. ✅ Port prototype tests to main test suite
3. ⚠️ Defer multi-term factoring to Sprint 12 (distribution cancellation alone achieves 39% reduction)
4. ✅ No blocking issues identified - ready to proceed

## Risk Assessment

**Overall Risk:** ✅ **LOW** - Safe to proceed with high confidence

- Performance regression: Low (57× faster than threshold)
- Incorrect factoring: Low (comprehensive test suite passes)
- Integration issues: Low (algorithm is self-contained)

## Quality Checks

- ✅ `make typecheck` - Pass
- ✅ `make lint` - Pass
- ✅ `make format` - Pass
- ✅ `make test` - Pass (1541 passed, 6 skipped, 1 xfailed)

## Related

- Sprint 11 Prep Task 5
- Verifies Unknowns 1.1, 1.7 in KNOWN_UNKNOWNS.md
- Part of Sprint 11 aggressive simplification feature

### CODE REVIEW

File: prototypes/aggressive_simplification/factoring_prototype.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+# Add src to path for imports
+sys.path.insert(0, str(Path(__file__).parent.parent.parent))
+
+from src.ir.ast import Binary, Const, Expr, VarRef
+
+if TYPE_CHECKING:
+    pass
+
+
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply prototypes/aggressive_simplification/factoring_prototype.py]: Fixed in 831495d. Removed the unused `TYPE_CHECKING` import and the empty conditional block. The import was not being used anywhere in the prototype code.
